core/container.go

```diff
@@ -21,72 +22,315 @@ type Container struct {
+		return nil, err
+	}
+
+	dirSt, dirRel, err := source.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	dirDef, err := dirSt.Marshal(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	payload.Mounts = append(payload.Mounts, ContainerMount{
+		Source:     dirDef.ToPB(),
+		SourcePath: dirRel,
+		Target:     target,
 	})
+
+	id, err := payload.Encode()
 	if err != nil {
 		return nil, err
 	}
 
-	return &Container{
-		ID: ContainerID(id),
-	}, nil
+	return &Container{ID: id}, nil
+}
+
+func (container *Container) ImageConfig(ctx context.Context) (specs.ImageConfig, error) {
+	payload, err := container.ID.decode()
+	if err != nil {
+		return specs.ImageConfig{}, err
+	}
+
+	return payload.Config, nil
+}
+
+func (container *Container) UpdateImageConfig(ctx context.Context, updateFn func(specs.ImageConfig) specs.ImageConfig) (*Container, error) {
+	payload, err := container.ID.decode()
+	if err != nil {
+		return nil, err
+	}
+
+	payload.Config = updateFn(payload.Config)
+
+	id, err := payload.Encode()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Container{ID: id}, nil
+}
+
+func (container *Container) Exec(ctx context.Context, gw bkgw.Client, platform specs.Platform, args []string, opts ContainerExecOpts) (*Container, error) {
+	payload, err := container.ID.decode()
+	if err != nil {
+		return nil, fmt.Errorf("decode id: %w", err)
+	}
+
+	cfg := payload.Config
+	mounts := payload.Mounts
+
+	shimSt, err := shim.Build(ctx, gw, platform)
+	if err != nil {
+		return nil, fmt.Errorf("build shim: %w", err)
+	}
+
+	runOpts := []llb.RunOption{
+		// run the command via the shim, hide shim behind custom name
+		llb.AddMount(shim.Path, shimSt, llb.SourcePath(shim.Path)),
+		llb.Args(append([]string{shim.Path}, args...)),
+		llb.WithCustomName(strings.Join(args, " ")),
+		llb.AddMount(metaMount, llb.Scratch()),
+	}
+
+	if cfg.WorkingDir != "" {
+		runOpts = append(runOpts, llb.Dir(cfg.WorkingDir))
+	}
+
+	for _, env := range cfg.Env {
+		name, val, ok := strings.Cut(env, "=")
+		if !ok {
+			// it's OK to not be OK
+			// we'll just set an empty env
+			_ = ok
+		}
+
+		runOpts = append(runOpts, llb.AddEnv(name, val))
+	}
+
+	for _, mnt := range mounts {
+		st, err := mnt.SourceState()
+		if err != nil {
+			return nil, fmt.Errorf("mount %s: %w", mnt.Target, err)
+		}
+
+		mountOpts := []llb.MountOption{}
+		if mnt.SourcePath != "" {
+			mountOpts = append(mountOpts, llb.SourcePath(mnt.SourcePath))
+		}
+
+		runOpts = append(runOpts, llb.AddMount(mnt.Target, st, mountOpts...))
+	}
+
+	st, err := payload.FSState()
+	if err != nil {
+		return nil, fmt.Errorf("fs state: %w", err)
+	}
+
+	execSt := st.Run(runOpts...)
+
+	execDef, err := execSt.Root().Marshal(ctx, llb.Platform(platform))
+	if err != nil {
+		return nil, fmt.Errorf("marshal root: %w", err)
+	}
+
+	// propagate any changes to the mounts to subsequent containers
+	for i, mnt := range mounts {
+		execMountDef, err := execSt.GetMount(mnt.Target).Marshal(ctx, llb.Platform(platform))
```

Comment:
If you specify multiple mounts at the same location, I *think* this will not error output but you'll start getting the wrong mount here.

This gets into all the corner cases around overlapping mounts that we are just ignoring for the first iterations here, so fixing this isn't a blocker at all, but worth a followup issue.

One solution is to get the modified mount state by calling `execSt.AddMount` instead of `GetMount`, but we may want to handle this situation entirely differently, so can figure it out later.

Reply:
I see - so the issue is that we always append to `containerIDPayload.Mounts` which almost does what the user wants, since later ones override, but `GetMount` always gets the first mount from the list, i.e. the one that is going to be mounted over by the later ones.

> One solution is to get the modified mount state by calling `execSt.AddMount` instead of `GetMount`, but we may want to handle this situation entirely differently, so can figure it out later.

I'm not sure I follow, but I guess another way could be to replace any mounts at or beneath the same target when adding a mount point, rather than just appending?

Reply:
I'll save the fix here for after #3217 is merged since it adds a bunch of other mount APIs that'll need to be updated. Thanks for noticing it!

Reply:
Yeah we need to decide what we want the behavior to be in overlapping mount cases, including once we support `withoutMount`. My default assumption is that we just mirror the behavior of how mounts work in Linux where if you make multiple mounts over the same point then they "stack" in that if you unmount one you will reveal the one below it.

In that case, the problem is that in this loop all the mounts at a given point will be set to the same one, so if you later unmount you will no longer have the mount you were expecting.

The suggestion to use `AddMount` is that in LLB instead of providing mounts as run options you can also do this:
```go
execState := llb.Image(...).Run(...)
mountA := execState.AddMount(foo, "/foo")
mountB := execState.AddMount(bar, "/bar")
```
in which case `mountA` and `mountB` will be the state of the mounts *after* the exec occurred. Basically a different way of accomplishing what `GetMount` accomplishes, but in such a way that I think this case of overlapping mounts will be handled.

Either way, yeah this will be important to get right someday for consistency but it's getting far beyond what we need to implement right now, saving for after #3217 sounds good to me.

Reply:
Oh nice, yeah that's a much cleaner fix! Just confirmed locally with a test I wrote for the other fix I attempted and it also makes it pass. I like the idea of just deferring to the platform's native mount layering semantics.

Here's the fix: https://github.com/dagger/dagger/pull/3217/commits/522b11f1e6771163b41c91029ef3877058df5f13